### PR TITLE
 Magento2 231 configurable products are updated

### PR DESCRIPTION
**Pull request by andre-kraus**
# Pull Request Template

## Description

Implemented update sync for configurable products + attribute mass update for products.

## Type of change

- [ ] Bug fix
- [x] New feature
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I updated the CHANGELOG.md

### CODE REVIEW

**Comment by augsteyer** [comment]
I'll check more tomorrow, but I get this:
```
) Shopgate\Connect\Test\Integration\Plugin\Catalog\Product\MassUpdateTestTest::testMassUpdatedForSimpleAndConfigurableProduct
Error: Call to a member function updateProduct() on null

/var/www/html/vendor/shopgate/connect-integration-magento2/src/Service/Catalog/Product.php:195
/var/www/html/vendor/shopgate/connect-integration-magento2/src/Plugin/Catalog/Product/MassUpdate.php:77
/var/www/html/vendor/shopgate/connect-integration-magento2/src/Test/Integration/Plugin/Catalog/Product/MassUpdateTest.php:207
```

File: src/Service/Catalog/Product.php [hunk 1]
```diff
@@ -145,19 +157,49 @@ private function shouldSynchronize(MageProduct $product)
...
+
+                $product->addData($attributeData);
+                $updates = $this->aggregateProductIdsForUpdate($product);
+
+                foreach ($updates as $productId => $dto) {
+                    $this->utility->initSdk()->getCatalogService()->updateProduct((string) $productId, $dto);
+                }
+            } catch (UnsupportedProductType $exception) {
+                continue;
+            }
```
Comment (augsteyer) [inline src/Service/Catalog/Product.php]: I'd recommend just checking sync for performance:
```php
try {
   $this->shouldSynchronize($product);
} catch (UnsupportedProductType $exception) {
   continue;
}
$product->addData($attributeData);
$updates = $this->aggregateProductIdsForUpdate($product);
 
foreach ($updates as $productId => $dto) {
    $this->utility->initSdk()->getCatalogService()->updateProduct((string) $productId, $dto);
}
```

File: src/Service/Catalog/Product.php [hunk 2]
```diff
@@ -171,6 +213,54 @@ public function update(MageProduct $product): void
...
+
+    /**
+     * @param MageProduct $product
+     *
+     * @return array
+     */
+    private function aggregateProductIds(MageProduct $product): array
+    {
+        $productIds = [
+            $product->getId()
```
Comment (augsteyer) [inline src/Service/Catalog/Product.php]:
```suggestion
            (string) $product->getId()
```

File: src/Service/Catalog/Product.php [hunk 3]
```diff
@@ -171,6 +213,54 @@ public function update(MageProduct $product): void
...
+            }
+        }
+
+        return $updatesToSend;
+    }
+
+    /**
+     * @param MageProduct $product
+     *
+     * @return array
```
Comment (augsteyer) [inline src/Service/Catalog/Product.php]:
```suggestion
     * @return string[]
```

File: src/Service/Catalog/Product.php [hunk 4]
```diff
@@ -171,6 +213,54 @@ public function update(MageProduct $product): void
     public function delete(MageProduct $product): void
     {
         $this->shouldSynchronize($product);
-        $this->utility->initSdk()->getCatalogService()->deleteProduct((string) $product->getId());
+
+        foreach ($this->aggregateProductIds($product) as $productId) {
+            $this->utility->initSdk()->getCatalogService()->deleteProduct((string) $productId);
```
Comment (augsteyer) [inline src/Service/Catalog/Product.php]: should be string, unless u are not a believer in ur code
```suggestion
            $this->utility->initSdk()->getCatalogService()->deleteProduct($productId);
```

File: src/Service/Catalog/Product.php [hunk 5]
```diff
@@ -145,19 +157,49 @@ private function shouldSynchronize(MageProduct $product)
...
-        }
-
         $this->shouldSynchronize($product);
-        $list = $this->productHelper->getProductDtoMap($product, ProductHelper::ACTION_UPDATE);
-        foreach ($list as $storeGroup => $dto) {
-            $this->utility->initSdk()->getCatalogService()->updateProduct((string) $product->getId(), $dto);
+        $updates = $this->aggregateProductIdsForUpdate($product);
+
+        foreach ($updates as $productId => $dto) {
+            $this->utility->initSdk()->getCatalogService()->updateProduct((string) $productId, $dto);
```
Comment (augsteyer) [inline src/Service/Catalog/Product.php]: make sure this is true on the `aggregateProductIdsForUpdate` end
```suggestion
            $this->utility->initSdk()->getCatalogService()->updateProduct($productId, $dto);
```

File: src/Service/Catalog/Product.php [hunk 6]
```diff
@@ -171,6 +213,54 @@ public function update(MageProduct $product): void
...
+        foreach ($this->aggregateProductIds($product) as $productId) {
+            $this->utility->initSdk()->getCatalogService()->deleteProduct((string) $productId);
+        }
+    }
+
+
+    /**
+     * @param MageProduct $product
+     *
+     * @return array
```
Comment (augsteyer) [inline src/Service/Catalog/Product.php]: what kind of array

File: src/Test/Integration/Plugin/Catalog/Product/MassUpdateTest.php [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+        $this->objectManager->get(MassUpdate::class)
+                            ->afterUpdateAttributes($subject, $result, $productIds, $attributeData, $storeId);
+    }
+
+    /**
+     * Injects mock utility into global object manager
+     * to prevent actual calls to the SDK
+     */
+    protected function setUp()
```
Comment (augsteyer) [inline src/Test/Integration/Plugin/Catalog/Product/MassUpdateTest.php]: oh, setUp on the bottom, auto formatter I am assuming?

File: src/Test/Integration/Plugin/Catalog/Product/MassUpdateTest.php [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+        $result = $this->getMockBuilder(Action::class)
+                       ->disableOriginalConstructor()
+                       ->setMethods([])
+                       ->getMock();
+
+        $productIds    = [23, 682];
+        $attributeData = ['description' => $description];
+        $storeId       = 1;
+
+        $this->objectManager->get(MassUpdate::class)
```
Comment (augsteyer) [inline src/Test/Integration/Plugin/Catalog/Product/MassUpdateTest.php]: oh wow, u are calling it directly. is there a way to call magento's frontend stuff?

File: src/Observer/Catalog/Product/Mass.php [hunk 9]
```diff
@@ -0,0 +1,76 @@
+<?php
```
[the remark this answers is not in the document]
Comment (andre-kraus) [reply src/Observer/Catalog/Product/Mass.php]: Mass delete is working out of the box as it is handled by the default delete observer. So I will go for MassUpdate

File: src/Test/Integration/Plugin/Catalog/Product/MassUpdateTest.php (hunk 7 above)
Comment (andre-kraus) [reply src/Test/Integration/Plugin/Catalog/Product/MassUpdateTest.php]: Yep, will move it up manually.

File: src/Test/Integration/Plugin/Catalog/Product/MassUpdateTest.php (hunk 8 above)
Comment (andre-kraus) [reply src/Test/Integration/Plugin/Catalog/Product/MassUpdateTest.php]: I just followed the approach, how magento is testing those plugins within integration tests.

File: src/Observer/Catalog/Product/Mass.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (andre-kraus) [reply src/Observer/Catalog/Product/Mass.php]: Hah, good catch with the missing event.xml modification. The observer is not used at all anymore. I am using an after plugin (Plugin/Catalog/Product/MassUpdate.php) and luckily it is already named perfectly.

File: src/Service/Catalog/Product.php [hunk 10]
```diff
@@ -145,19 +159,49 @@ private function shouldSynchronize(MageProduct $product)
...
+                $this->shouldSynchronize($product);
+            } catch (UnsupportedProductType $exception) {
+                continue;
+            }
+
+            $product->addData($attributeData);
+            $updates = $this->aggregateProductIdsForUpdate($product);
+
+            foreach ($updates as $productId => $dto) {
+                $this->utility->initSdk()->getCatalogService()->updateProduct((string) $productId, $dto);
```
Comment (augsteyer) [inline src/Service/Catalog/Product.php]: missed one
```suggestion
                $this->utility->initSdk()->getCatalogService()->updateProduct($productId, $dto);
```

File: src/Test/Integration/Plugin/Catalog/Product/MassUpdateTest.php [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+use PHPUnit\Framework\MockObject\MockObject;
+use PHPUnit\Framework\TestCase;
+use Shopgate\Connect\Plugin\Catalog\Product\MassUpdate;
+use Shopgate\Connect\Service\Catalog\Product as ProductService;
+use Shopgate\Connect\Service\Utility;
+use Shopgate\ConnectSdk\Dto\Catalog\Product;
+use Shopgate\ConnectSdk\Service\Catalog as CatalogService;
+use Shopgate\ConnectSdk\ShopgateSdk;
+
+class MassUpdateTestTest extends TestCase
```
Comment (augsteyer) [inline src/Test/Integration/Plugin/Catalog/Product/MassUpdateTest.php]: u are not isolating these tests, check other tests

**Comment by augsteyer** [comment]
may need to restart build to see if it passes with automated tests now. The sample DB is still not quite the same even though I properly install it on first run.

**Comment by augsteyer** [comment]
I recommend using SKU's and just do `productRepository->get($sku)`, this will ensure we are loading the right product no matter what version of sample data. A better idea is to start moving off of the sample database completely by creating the products we need, this will speed up our tests in the long run. [Here](https://github.com/shopgate/cart-integration-magento2-export/pull/28/commits/c354a8e439bb27a1e45eee5005cdaf53b34bc73b#diff-da5d566efbe253bab9a90c0337675c57R30) is an example in the export PR. The creation of products is much faster than website/store creation ^_^

**Pull request merged**